src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -600,6 +611,155 @@ fromChunks cs FinalChunk{..} = CompactIndex {
+-- serialised compact index, with no leading or trailing space.
+-- It is directly used as the backing memory for the compact index.
+--
+-- Also note that the implementation assumes a little-endian system.
+--
+-- __NOTE__: Currently does not perform bounds checks. Malformed input can
+-- trigger invalid memory access!
+fromSBS :: ShortByteString -> Either String (NumEntries, CompactIndex)
+fromSBS (SBS ba') = do
+    let ba = ByteArray ba'
+    when (mod8 (sizeofByteArray ba) /= 0) $ Left "Length is not multiple of 64 bit"
+
+    let arr64 = P.PrimArray ba' :: P.PrimArray Word64
+    let len64 = P.sizeofPrimArray arr64
+
+    when (len64 < 3) $ Left "Doesn't contain size information"
+    let ciRangeFinderPrecision = fromIntegral (P.indexPrimArray arr64 (len64 - 3))
+    let numPages = fromIntegral (P.indexPrimArray arr64 (len64 - 2))
+    let numEntries = fromIntegral (P.indexPrimArray arr64 (len64 - 1))
+    let numRanges = 2 ^ ciRangeFinderPrecision + 1
+
+    -- offsets in 32 bits
+    let (offset2_32, ciPrimary) = getVec32 ba 0 numPages
+    let (offset3_32, ciRangeFinder) = getVec32 ba offset2_32 numRanges
+    -- offsets in 64 bits
+    let offset3 = divUp2 offset3_32
+    let (offset4, ciClashes) = getBitVec ba offset3 numPages
+    let (offset5, ciLargerThanPage) = getBitVec ba offset4 numPages
+    let ciTieBreaker = getTieBreaker ba offset5
+
+    return (NumEntries numEntries, CompactIndex {..})
+
+type Offset32 = Int
+type Offset64 = Int
+
+getVec32 :: ByteArray -> Offset32 -> Int -> (Offset32, VU.Vector Word32)
+getVec32 ba offset32 numEntries = (offset32 + numEntries, vec)
+  where
+    vec = VU.V_Word32 (VP.Vector offset32 numEntries ba)
+
+getBitVec :: ByteArray -> Offset64 -> Int -> (Offset64, VU.Vector Bit)
+getBitVec ba offset64 numEntries = (offset64', vec)
+  where
+    vec = BitVec (mul64 offset64) numEntries ba
+    offset64' = offset64 + divUp64 numEntries
+
+getTieBreaker :: ByteArray -> Offset64 -> Map SerialisedKey PageNr
+getTieBreaker ba = \offset64 ->
+    let size = fromIntegral (indexByteArray ba offset64 :: Word64)
+    in Map.fromList $ take size $ unfoldr (Just . getEntry) (offset64 + 1)
+  where
+    getEntry :: Offset64 -> ((SerialisedKey, PageNr), Offset64)
+    getEntry offset64 =
+        let offset32 = mul2 offset64
+            pageNr = fromIntegral (indexByteArray ba offset32 :: Word32)
+            keyLen8 = fromIntegral (indexByteArray ba (offset32 + 1) :: Word32)
+            keyOffset8 = mul8 (offset64 + 1)
+            -- We avoid retaining references to the bytearray.
+            -- It will stay alive as long as the compact index anyway, so we
+            -- probably don't need the 'VP.force', but we expect very few keys
+            -- in the tie breaker, so it is cheap to do and we don't have to
+            -- worry about the issue anymore.
+            key = SerialisedKey' (VP.force (VP.Vector keyOffset8 keyLen8 ba))
```

Comment:
You should probably `!` all the map entries here (`key`, `pageNr` etc). Yes they will all get forced by the `Map.fromList` later on, but that's a bunch of unnecessary thunks.